### PR TITLE
feat: implement bats test runner

### PR DESCRIPTION
**Pull request by thesayyn**
### Type of change

- New feature or functionality (change which adds functionality)
- Relevant documentation has been updated

### Test plan

- New test cases added

### CODE REVIEW

File: lib/repositories.bzl [hunk 1]
```diff
@@ -103,6 +104,81 @@ def register_tar_toolchains(name = DEFAULT_TAR_REPOSITORY, register = True):
         user_repository_name = name,
     )
 
+DEFAULT_BATS_REPOSITORY = "bats"
+
+DEFAULT_BATS_CORE_VERSION = "v1.10.0"
```
Comment (jbedard) [inline lib/repositories.bzl:109]: We don't want to make all these the latest version?

**Comment by jbedard** [comment]
Did you try this out in silo?

**Comment by thesayyn** [comment]
> Did you try this out in silo?

Not directly, just added usages there as tests here.

File: lib/repositories.bzl (hunk 1 above)
Comment (thesayyn) [reply lib/repositories.bzl:109]: these are all latest versions as of today.

File: lib/private/bats.bzl [hunk 2]
```diff
@@ -0,0 +1,90 @@
+"bats_test"
+
+load("@aspect_bazel_lib//lib:expand_make_vars.bzl", "expand_locations", "expand_variables")
+load("@aspect_bazel_lib//lib:paths.bzl", "BASH_RLOCATION_FUNCTION", "to_rlocation_path")
+
+_RUNNER_TMPL = """#!/usr/bin/env bash
+
+{BASH_RLOCATION_FUNCTION}
+
+# set -o errexit -o nounset -o pipefail
```
Comment (thesayyn) [inline lib/private/bats.bzl:10]: revert

File: lib/private/bats_toolchain.bzl [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        "libraries": "bats helper libraries",
+    },
+)
+
+def _bats_toolchain_impl(ctx):
+    core = ctx.file.core
+
+    # Make the $(BATS_BIN) variable available in places like genrules.
+    # See https://docs.bazel.build/versions/main/be/make-variables.html#custom_variables
+    template_variables = platform_common.TemplateVariableInfo({})
```
Comment (kormide) [inline lib/private/bats_toolchain.bzl:76]: Is there a template variable you forgot to add here?

File: lib/extensions.bzl [hunk 4]
```diff
@@ -104,5 +116,9 @@ toolchains = module_extension(
         "coreutils": tag_class(attrs = {"name": attr.string(default = DEFAULT_COREUTILS_REPOSITORY), "version": attr.string(default = DEFAULT_COREUTILS_VERSION)}),
         "tar": tag_class(attrs = {"name": attr.string(default = DEFAULT_TAR_REPOSITORY)}),
         "expand_template": tag_class(attrs = {"name": attr.string(default = DEFAULT_EXPAND_TEMPLATE_REPOSITORY)}),
+        "bats": tag_class(attrs = {
+            "name": attr.string(default = DEFAULT_BATS_REPOSITORY),
+            "core_version": attr.string(default = DEFAULT_BATS_CORE_VERSION),
```
Comment (kormide) [inline lib/extensions.bzl:121]: Want to add attributes for the versions of other libraries?

Comment (thesayyn) [reply lib/extensions.bzl:121]: Yeah, but I don't want to significantly change this function...

File: lib/private/bats_toolchain.bzl (hunk 3 above)
Comment (thesayyn) [reply lib/private/bats_toolchain.bzl:76]: Not really, running BATS is little complex and wouldn't fit we'll in a genrule.

Comment (thesayyn) [reply lib/private/bats_toolchain.bzl:76]: I'll leave that for future.

File: lib/extensions.bzl (hunk 4 above)
Comment (kormide) [reply lib/extensions.bzl:121]: I think it should just match what you can do with the WORKSPACE macro.

File: lib/private/bats_toolchain.bzl (hunk 3 above)
Comment (kormide) [reply lib/private/bats_toolchain.bzl:76]: I agree that it doesn't really fit. Can probably just remove this block then.

**Comment by alexeagle** [comment]
@jmmv do you have any input about this as a comparison with https://github.com/jmmv/rules_shtk/ ?

File: lib/extensions.bzl (hunk 4 above)
Comment (thesayyn) [reply lib/extensions.bzl:121]: can i leave that as a follow up?

Comment (thesayyn) [reply lib/extensions.bzl:121]: I need this for rules_oci and silo both is not using bzlmod ATM.

**Review by kormide — APPROVED** [review]
🚀

**Comment by thesayyn** [comment]
before i merge, let me try this on rules_oci.

File: lib/repositories.bzl (hunk 1 above)
Comment (jbedard) [reply lib/repositories.bzl:109]: I mean doing `= BATS_CORE_VERSIONS[-1]` or whatever the syntax would be...

**Pull request merged**